doc/adrs/086_provide_the_variables_available_for_each_operation.adoc

```diff
@@ -0,0 +1,170 @@
+ADR-086 - Provide the variables available for each operation
+
+== Context
+
+We need to know the variables available for an operation to document them, provide content assist and validate operations to help specifiers.
+
+All operations may have different variables.
+Given the way a representation description is organized, the number of variables and their type may change.
+We need to give access to the list of the variables in:
+
+- Our documentation
+- The Java API
+- The details view for the View DSL
+
+We need to keep in mind that a single operation implemented by the Java API may be transformed into multiple operations in the view DSL with different variables.
+A simple Java tool handler may become a complex set of model operations with several expressions which may come with specific variables.
+
+This is both a very important topic and a very complex one.
+We can't fulfill every requirements on the first try to we will have to try to cover as much as possible with a small investment.
+
+== Decision
+
+=== Define operations
+
+We will list all the operations in the application.
+An operation may be implemented by one or many Java handlers or expressions.
+It is a simple key used to attach variables to Java code or expressions.
+We will thus need to map operations with the way they are interpreted (Java code or expressions).
+
+For that, we will start listing all the operations available, for example in the diagram representation:
+
+[source,java]
+----
+public interface IOperationProvider {
+  List<String> getOperations();
+}
+----
+
+[source,java]
+----
+public class DiagramOperationProvider implements IOperationProvider {
+  @Override
+  public <String> getOperations() {
+    return List.of(
+      "Diagram#canCreate",
+      "Diagram#label",
+      "Diagram#targetObjectId",
+      "Node#semanticElements",
+      "Edge#shouldRender"
+    );
+  }
+}
+----
+
+=== Define variables
+
+Once the operations are defined, we can define the variables which can be available for an operation.
+
+[source,java]
+----
+public interface IVariableProvider {
+  List<Variable> getVariables(String operation);
+}
+----
+
+[source,java]
+----
+public record Variable(String name, Class<?> type, String documentation) {}
```

Comment:
It may be needed to support multiple types for a given variable, see the signature of `org.eclipse.acceleo.query.runtime.IQueryCompletionEngine.getCompletion(String, int, Map<String, Set<IType>>)`.